src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.cpp

```diff
@@ -0,0 +1,163 @@
+#include "Utilities/Gsl.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+
+/// \cond
+namespace GeneralizedHarmonic {
+template <size_t VolumeDim, typename Frame, typename DataType>
+void spatial_ricci_tensor(
+    const gsl::not_null<tnsr::ii<DataType, VolumeDim, Frame>*> ricci,
+    const tnsr::iaa<DataType, VolumeDim, Frame>& phi,
+    const tnsr::ijaa<DataType, VolumeDim, Frame>& deriv_phi,
+    const tnsr::II<DataType, VolumeDim, Frame>&
+        inverse_spatial_metric) noexcept {
+  if (UNLIKELY(get_size(get<0, 0>(*ricci)) != get_size(get<0, 0, 0>(phi)))) {
+    *ricci = tnsr::ii<DataType, VolumeDim, Frame>(get_size(get<0, 0, 0>(phi)));
+  }
+
+  TempBuffer<tmpl::list<::Tags::TempIjj<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempijk<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<1, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<0, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<1, VolumeDim, Frame, DataType>>>
+      local_buffer(get_size(get<0, 0>(inverse_spatial_metric)));
+
+  // New variable to avoid recomputing in nested loops
+  auto& phi_uplolo =
+      get<::Tags::TempIjj<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+        phi_uplolo.get(k, i, j) = 0.;
+        for (size_t l = 0; l < VolumeDim; ++l) {
+          phi_uplolo.get(k, i, j) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(l, 1 + i, 1 + j);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& phi_loloup =
+      get<::Tags::Tempijk<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = 0; j < VolumeDim; ++j) {
+        phi_loloup.get(i, j, k) = 0.;
+        for (size_t l = 0; l < VolumeDim; ++l) {
+          phi_loloup.get(i, j, k) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(i, 1 + j, 1 + l);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& phi_trace1 =
+      get<::Tags::Tempi<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    phi_trace1.get(i) = 0.;
+    for (size_t j = 0; j < VolumeDim; ++j) {
+      phi_trace1.get(i) += phi_loloup.get(i, j, j);
+    }
+  }
+  auto& phi_trace2 =
+      get<::Tags::Tempi<1, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    phi_trace2.get(i) = 0.;
+    for (size_t j = 0; j < VolumeDim; ++j) {
+      phi_trace2.get(i) += phi_uplolo.get(j, j, i);
+    }
+  }
+
+  auto& phi_trace1_up =
+      get<::Tags::TempI<0, VolumeDim, Frame, DataType>>(local_buffer);
+  auto& phi_trace2_up =
```

Comment:
update these names consistent with the above suggested changes

Reply:
done